### PR TITLE
feat: Update VM (use existing PIP)

### LINKED ISSUE #2702: [AVM Module Issue]: PIP zones breaks network connectivity when not using zones.
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Issue Type?

Bug

### Module Name

avm/res/compute/virtual-machine

### (Optional) Module Version

0.5.3

### Description

When trying to create a vm with zone = 0 and an availability set the following error occurs while trying to update an existing publicIP:
"pip has an existing availability zone constraint NoZone and the request has availability zone constraint 1, 2, 3, which do not match. Zones cannot be added/updated/removed once the resource is created. The resource cannot be updated from regional to zonal or vice-versa."

However if a new public ip is used the zone value seems to be incorrectly set to an array [1,2,3] resulting in broken network connectivity on the VM. Manually replacing the public ip fixes the connectivity issues. Exporting the Public IP template shows the odd zone value somehow being set to the full array.  

This looks to stem from lines 64 - 67 of https://github.com/Azure/bicep-registry-modules/blob/avm/res/compute/virtual-machine/0.5.3/avm/res/compute/virtual-machine/modules/nic-configuration.bicep

Zones is optional on public ips and if the zone isn't provided it should not be set. Those lines seem to be setting it to all allowed values.

Sample of the exported template:
```
{
    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
        "publicIPAddresses_vmname_pip_name": {
            "defaultValue": "vnamne-pip",
            "type": "String"
        }
    },
    "variables": {},
    "resources": [
        {
            "type": "Microsoft.Network/publicIPAddresses",
            "apiVersion": "2023-11-01",
            "name": "[parameters('publicIPAddresses_vmname_pip_name')]",
            "location": "eastus",
            "sku": {
                "name": "Standard",
                "tier": "Regional"
            },
            "zones": [
                "1",
                "2",
                "3"
            ],
            "properties": {
                "ipAddress": "1.1.1.1",
                "publicIPAddressVersion": "IPv4",
                "publicIPAllocationMethod": "Static",
                "idleTimeoutInMinutes": 4,
                "dnsSettings": {
                    "domainNameLabel": "vmname",
                    "fqdn": "vmname.eastus.cloudapp.azure.com"
                },
                "ipTags": [],
                "ddosSettings": {
                    "protectionMode": "VirtualNetworkInherited"
                }
            }
        }
    ]
}
```

### (Optional) Correlation Id

_No response_

### PR DESCRIPTION
**Pull request by rahalan**
closes #2702 

An existing PIP can be provided.

run (except NVidia): [avm.res.compute.virtual-machine](https://github.com/rahalan/bicep-registry-modules/actions/workflows/avm.res.compute.virtual-machine.yml)

### CODE REVIEW

File: avm/res/compute/virtual-machine/README.md [hunk 1]
```diff
@@ -4423,11 +4421,13 @@ This section gives you an overview of all local-referenced module files (i.e., o
...
 | Reference | Type |
 | :-- | :-- |
-| `br/public:avm/res/network/network-interface:0.2.4` | Remote reference |
-| `br/public:avm/res/network/public-ip-address:0.4.1` | Remote reference |
+| `br/public:avm/res/network/network-interface:0.4.0` | Remote reference |
+| `br/public:avm/res/network/public-ip-address:0.6.0` | Remote reference |
 
 ## Notes
 
+Inside the `nicConfigurations` section and there inside the `ipConfigurations`, a `pipConfiguration` can be defined. For a new puplic IP address, the naming can either be set with the `name` or the `publicIpNameSuffix`. Per default a newly created PIP will have its `zones` parameter set to `[1,2,3]`. You can override it, for example with `[]`. If an existing PIP should be used, only set the `publicIPAddressResourceId`.
```
Comment (eriqua) [inline avm/res/compute/virtual-machine/README.md:4429]: Given the new functionality, I'd suggest to raise the minor version

File: avm/res/compute/virtual-machine/tests/e2e/windows.max/main.test.bicep
Comment (eriqua) [inline avm/res/compute/virtual-machine/tests/e2e/windows.max/main.test.bicep:1]: General comment, not related to this test file in particular:
I see multiple e2e tests failing. I know you are on it. Please let me know here once ready (all but nvidia of course)

File: avm/res/compute/virtual-machine/modules/nic-configuration.bicep [hunk 2]
```diff
@@ -27,51 +27,40 @@ param diagnosticSettings diagnosticSettingType
 @description('Optional. Array of role assignments to create.')
 param roleAssignments roleAssignmentType
 
-module networkInterface_publicIPAddresses 'br/public:avm/res/network/public-ip-address:0.4.1' = [
-  for (ipConfiguration, index) in ipConfigurations: if (contains(ipConfiguration, 'pipConfiguration')) {
+var publicIpNameSuffix = '-pip-01'
```
Comment (eriqua) [inline avm/res/compute/virtual-machine/modules/nic-configuration.bicep:30]: Not 100% sure about this hardcoded value here. Shouldn't this come as an input parameter? At least we should document somewhere that if neither name nor suffix is provided this default value will be used. Thoughts?

**Comment by ReneHezser** [comment]
@rahalan 
We have added examples for Bicep parameter files to the Readme. This has been applied to all published modules but needs to be done for PRs as well. Can you please run the Set-AVMModule utility as detailed [here](https://azure.github.io/Azure-Verified-Modules/contributing/bicep/bicep-contribution-flow/generate-bicep-module-files/). It is required for the validation pipeline to succeed and the contribution to be published.

Please reach out if any support is needed.

File: avm/res/compute/virtual-machine/README.md (hunk 1 above)
Comment (rahalan) [reply avm/res/compute/virtual-machine/README.md:5751]: Done

File: avm/res/compute/virtual-machine/tests/e2e/windows.max/main.test.bicep
Comment (rahalan) [reply avm/res/compute/virtual-machine/tests/e2e/windows.max/main.test.bicep:1]: Done, finally. 
image

File: avm/res/compute/virtual-machine/modules/nic-configuration.bicep (hunk 2 above)
Comment (rahalan) [reply avm/res/compute/virtual-machine/modules/nic-configuration.bicep]: removed

**Review by eriqua — COMMENTED** [review]
@rahalan approving. 

Just a suggestion for a next PR: it would be useful to adopt UDT for  the nicConfigurations object. This would help documenting the object property descriptions and in general it would be very helpful for module consumers.

**Review by eriqua — APPROVED** [review]
🆗

**Pull request merged**